pypeit/slittrace.py

```diff
@@ -721,140 +723,296 @@ def slit_spat_pos(left, right, nspat):
+                                     (sobjs.MASKDEF_OBJNAME != 'SERENDIP')
+                if np.any(target_obj_in_slit):
+                    continue
+
+            # Object index
+            oidx = np.where(self.maskdef_designtab['SLITID'].data == self.maskdef_id[islit])[0][0]
+            #
+            # Do it
+            SPAT_PIXPOS = float((expected_objpos[oidx]-self.mask_median_off) + slits_left[specmid, islit])
+
+            # TODO -- DO THIS RIGHT FOR SLITS (LIKELY) OFF THE DETECTOR/SLIT
+            #  If we keep what follows, probably should add some tolerance to be off the edge
+            #  Otherwise things break in skysub
+            if (SPAT_PIXPOS > slits_right[specmid, islit]) or (SPAT_PIXPOS < slits_left[specmid, islit]):
+                msgs.warn("Targeted object is off the detector")
+                continue
+
+            # Generate a new specobj
+            specobj_dict = {'SLITID': self.spat_id[islit], # Confirm this
+                            'DET': self.det, 
+                            'OBJTYPE': 'science',  # Confirm this is ok
+                            'PYPELINE': self.pypeline}
+            thisobj = specobj.SpecObj(**specobj_dict)
+
+            # Fill me up
+            if sobjs.nobj == 0:
+                # This uses slit edge trace
+                xoff = SPAT_PIXPOS - slit_cen[specmid, islit]
+                thisobj.TRACE_SPAT = slit_cen[:, islit] + xoff
+                thisobj.SPAT_PIXPOS = SPAT_PIXPOS
+                thisobj.SPAT_FRACPOS = (SPAT_PIXPOS - slits_left[specmid, islit]) / \
+                                       (slits_right[specmid, islit]-slits_left[specmid, islit])
+                thisobj.trace_spec = np.arange(slits_left.shape[0])
+                # OBJID
+                thisobj.OBJID = 1
+            else:
+                # This uses an object
+                idx_nearest = np.argmin(np.abs(sobjs.SPAT_PIXPOS-SPAT_PIXPOS))
+                specmid = sobjs[idx_nearest].TRACE_SPAT.size//2
+                xoff = SPAT_PIXPOS - sobjs[idx_nearest].TRACE_SPAT[specmid]
+                thisobj.TRACE_SPAT = sobjs[idx_nearest].TRACE_SPAT + xoff
+                thisobj.SPAT_PIXPOS = SPAT_PIXPOS
+                thisobj.trace_spec = sobjs[idx_nearest].trace_spec
+                thisobj.SPAT_FRACPOS = (SPAT_PIXPOS - slits_left[specmid, islit]) / \
+                                       (slits_right[specmid, islit]-slits_left[specmid, islit])
+                # OBJID
+                any_obj_in_slit = sobjs.MASKDEF_ID == self.maskdef_id[islit]
+                if np.any(any_obj_in_slit):
+                    thisobj.OBJID = np.max(sobjs[any_obj_in_slit].OBJID) + 1
+                else:
+                    thisobj.OBJID = 1
+
+            # FWHM
+            thisobj.FWHM = fwhm  # pixels
+            thisobj.maskwidth = 4. * fwhm  # matches objfind() in extract.py
+            # Finishing up
+            thisobj.set_name()
+            # Mask info
```

Comment:
Here for example you could just test whether the the mask design information is available before setting these fields.